### ISSUE TITLE
String values in ExpressionAttributeValue do not support emoji

### ISSUE THREAD

**Issue by a90120411** [aws/aws-sdk-js-v3#3559]
## Describe the bug

When the string value in ExpressionAttributeValues contains emoji, `UpdateCommand` and `PutCommand` can't save data normally.

#### SDK version number
`@aws-sdk/lib-dynamodb@3.17.0`

#### Is the issue in the browser/Node.js/ReactNative?
Browser

#### Details of the browser/Node.js/ReactNative version
Chrome: 100.0.4896.127

## Steps to reproduce
```javascript

function initDynamoDBClient() {
  dynamoDBClient = new DynamoDBClient({
    region: "ap-east-1",
    credentials: {
      accessKeyId: AWSAccessKeyId,
      secretAccessKey: AWSSecretAccessKey,
    },
    endpoint: "https://dynamodb.ap-east-1.amazonaws.com",
  });
  const marshallOptions = {
    convertEmptyValues: false,
    removeUndefinedValues: false,
    convertClassInstanceToMap: false
  };
  const unmarshallOptions = {
    wrapNumbers: false
  };
  const translateConfig = { marshallOptions, unmarshallOptions };
  ddbDocClient = DynamoDBDocumentClient.from(dynamoDBClient, translateConfig);
  console.log("initDynamoDBClient finished");
}

async function updateHistoryItem(item) {
  const params = {
    TableName: tbName,
    Key: {
      url: item.url,
    },
    ExpressionAttributeValues: {
      ":read_state": item.read_state || 0,
      ":title": item.title,   //title is a string , include emoji.
    },
    UpdateExpression: `SET 
    title = :title,
    read_state = :read_state`,
  };
  return updateItemFromDynamoDB(params);
}

async function updateItemFromDynamoDB(params) {
  return sendCommandToDynamoDB(new UpdateCommand(params));
}

async function putItemFromDynamoDB(params) {
  return sendCommandToDynamoDB(new PutCommand(params));
}

async function sendCommandToDynamoDB(command) {
  try {
    const data = await ddbDocClient.send(command);
    console.log("AWS DynamoDB send command -- succeeded");
    return data;
  } catch (err) {    
    console.error(
      "AWS DynamoDB send command -- failed:" + "\n" + JSON.stringify(err)
    );
    return null;
  }
}

```

#### Observed behavior
Exception was thrown

```
failed: {"__type":"com.amazon.coral.service#InvalidSignatureException","name":"InvalidSignatureException","$fault":"client","$metadata":{"httpStatusCode":400,"requestId":"ML8R85HMGCI4FL4KQA4NKHPG5VVV4KQNSO5AEMVJF66Q9ASUAAJG","attempts":3,"totalRetryDelay":477}}

InvalidSignatureException: The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.
```

#### Expected behavior

The string value containing the emoji was successfully saved or provide a function for encoding.

**Comment by RanVaknin** [comment]
@a90120411 Hi,

Im trying to replicate your code but it's missing a lot of things like the correct import statements, variable declarations.
and also it would be very helpful if you have the schema code you used to set up the Ddb table with.

Thanks!

**Comment by a90120411** [comment]
> @a90120411 Hi,
> 
> Im trying to replicate your code but it's missing a lot of things like the correct import statements, variable declarations. and also it would be very helpful if you have the schema code you used to set up the Ddb table with.
> 
> Thanks!

Sorry, I added these information.

```javascript

const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
import {
  DynamoDBDocumentClient,
  PutCommand,
  GetCommand,
  UpdateCommand,
} from "@aws-sdk/lib-dynamodb";

let dynamoDBClient = null;
let ddbDocClient = null;
```

**Comment by RanVaknin** [comment]
Hi @a90120411,

I am not able to replicate your error because I'm missing two things:
1. your schema and table creation code.
2. the `item` object you're feeding your function

I suspect this error is coming from the way you are referring to the values in your `item` object, but again this is only an assumption since I don't have your full code. 
Please take a look at this [Stack Overflow issue](https://stackoverflow.com/questions/66591418/aws-nodejs-sdk-v3-dynamodb-updateitem-typeerror-cannot-read-property-0-of-u). I know that the original poster did not experience the same issue as you but you might be able to draw some useful information from the solution.

Despite the missing information, I was able to :
- [Create a ddb Table](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/dynamodb-examples-using-tables.html)
- [Putting an item in the table](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/dynamodb-example-table-read-write.html#dynamodb-example-table-read-write-writing-an-item) with the following fields:
```
Partition key: Season (Number): 3
Sort key: Episode (Number): 10
Title (String): "Benihanna Christmas"
Rank (String): "⭐⭐⭐⭐"
```
<img width="600" alt="image" src="https://user-images.githubusercontent.com/50976344/171277498-2ff5b8f8-ffe7-4b24-95d1-8e85627294f3.png">


- [Updating the same item](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/dynamodb-example-table-read-write.html#dynamodb-example-table-update-an-item) by identifying it using partition and sort key, and passing the new updated value (five stars instead of 4):
```
Rank (String): "⭐⭐⭐⭐⭐"
```
[
<img width="600" alt="image" src="https://user-images.githubusercontent.com/50976344/171277285-0d5a7edd-23ab-4616-9bdf-90d9bf3c0931.png">
](url)

**Conclusion:**
DynamoDB's `UpdateCommand` and `PutCommand` do support emojis in `ExpressionAttributeValues`.
Please follow the tutorial I linked in each of my steps. 

Let me know if you have any other issues.

**Comment by a90120411** [comment]
I will add some info.

**Comment by RanVaknin** [comment]
@a90120411 
Have you tried following the steps I've mentioned in my walkthrough?

**Comment by a90120411** [comment]
@RanVaknin  
Thanks for the information, unfortunately it didn't solve the problem.

I created a test table: tb_demo, It contains two fields:  id:String(key) and  title:String.
Execute the following code( I use webpack to package and run in the browser):
If I remove the emoji symbol in the title value, everything works fine.

This problem only happens in browser environment. 
It works fine in Node.js environment and PartiQL editor.

```javascript

const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const {
  DynamoDBDocumentClient,
  PutCommand,
  GetCommand,
} = require("@aws-sdk/lib-dynamodb");

const tbName = "tb_demo";
let dynamoDBClient = null;
let ddbDocClient = null;

function initDynamoDBClient() {
  dynamoDBClient = new DynamoDBClient({
    region: "ap-east-1",
    credentials: {
      accessKeyId: "XXXXXXX",
      secretAccessKey: "XXXXXX",
    },
    endpoint: "https://dynamodb.ap-east-1.amazonaws.com",
  });
  const marshallOptions = {    
    convertEmptyValues: false,    
    removeUndefinedValues: false,
    convertClassInstanceToMap: false,
  };
  const unmarshallOptions = {    
    wrapNumbers: false,
  };
  const translateConfig = { marshallOptions, unmarshallOptions };
  ddbDocClient = DynamoDBDocumentClient.from(dynamoDBClient, translateConfig);
  console.log("initDynamoDBClient finished");
}

async function putHistoryItem(item) {
  const params = {
    TableName: tbName,
    Item: item,
  };
  return putItemFromDynamoDB(params);
}

async function getDetailHistoryItem(id) {
  const params = {
    TableName: tbName,
    Key: {
      id: id,
    },
  };
  return getItemFromDynamoDB(params);
}

async function getItemFromDynamoDB(params) {
  return sendCommandToDynamoDB(new GetCommand(params));
}

async function putItemFromDynamoDB(params) {
  return sendCommandToDynamoDB(new PutCommand(params));
}

async function sendCommandToDynamoDB(command) {
  try {
    const data = await ddbDocClient.send(command);
    console.log("AWS DynamoDB send command -- succeeded");
    return data;
  } catch (err) {
    console.error(err);
    console.error(
      "AWS DynamoDB send command -- failed:" + "\n" + JSON.stringify(err)
    );
    return null;
  }
}
//run
initDynamoDBClient();
putHistoryItem( {id: "1", Title: "😄Emoji"});
getDetailHistoryItem("1").then((data) => {
  console.log(data);
});

```

Exception information:

```
serviceWorker.js:27559 InvalidSignatureException: The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.
    at file:///Users/source/frontend/WebHistoryTool/aws-dynamodb/dist/serviceWorker.js:14125:68
    at step (file:///Users/source/frontend/aws-dynamodb/dist/serviceWorker.js:694:23)
    at Object.next (file:///Users/source/frontend/aws-dynamodb/dist/serviceWorker.js:675:53)
    at fulfilled (file:///Users/source/frontend/aws-dynamodb/dist/serviceWorker.js:665:58)
sendCommandToDynamoDB @ serviceWorker.js:27559
await in sendCommandToDynamoDB (async)
putItemFromDynamoDB @ serviceWorker.js:27545
putHistoryItem @ serviceWorker.js:27510
(anonymous) @ serviceWorker.js:27493
(anonymous) @ serviceWorker.js:27576
(anonymous) @ serviceWorker.js:27578

serviceWorker.js:27560 AWS DynamoDB send command -- failed:
{"__type":"com.amazon.coral.service#InvalidSignatureException","name":"InvalidSignatureException","$fault":"client","$metadata":{"httpStatusCode":400,"requestId":"GJOQ4811OATTBTAHBM75R0F9GRVV4KQNSO5AEMVJF66Q9ASUAAJG","attempts":3,"totalRetryDelay":326}}
```

**Comment by RanVaknin** [comment]
@a90120411 , 

Thanks for helping me reproduce this issue. This is incredibly odd! It only breaks when bundled with Webpack but only for certain emojis. Requests with  `⭐`emojis are sent fine, but the `😄` emoji that you used was not supported. It might be an encoding issue. I assigned the issue to one of the developers on the team but since this is an edge case this would probably take low priority. 

Please consider this hacky solution to unblock you while we work on a fix.  One of the things you could do is encode the emoji, and when pulling data from AWS decode the emoji to display correctly in your application. Something like:

_putItem.js_
```
//initialize client, client.send(putItem....)

ExpressionAttributeValues: {
        ":rank": { S: "utf8.encode("😄") } // would save it as "8J+YhA=="
    }
```

_getItem.js_
```
// const item = await getItem(key)

console.log(utf8.decode(item.rank) // would turn into "😄"
```

**Comment by RanVaknin** [comment]
@trivikr ,
I have confirmed this is not a service related issue by doing the same action from the AWS CLI:

```
aws dynamodb update-item \
    --table-name "TEST_TABLE" \
    --key file://test_folder/Key.json \
    --update-expression "SET #rank = :rank" \
    --expression-attribute-names file://test_folder/ExpressionAttributeNames.json \
    --expression-attribute-values file://test_folder/ExpressionAttributeValues.json \
    --region us-east-1 \
    
```

This is a concise version needed to repro the issue.
```
// ====> needs to be bundled with Webpack <====

import { UpdateItemCommand } from "@aws-sdk/client-dynamodb";
import { ddbClient } from "../libs/ddbClient.js";

export const params = {
    TableName: "TEST_TABLE",
    Key: {
        Season: { N: "3" } ,
        Episode: { N: "10"} ,
    },
    ExpressionAttributeNames:{
        "#rank": "Rank"
    },
    UpdateExpression: "set #rank = :rank",
    ExpressionAttributeValues: {
        ":rank": { S: "😄" }
    },
    ReturnValues: "ALL_NEW"
};
export const run = async () => {
    try {
        const data = await ddbClient.send(new UpdateItemCommand(params));
        console.log(data);
        return data;
    } catch (err) {
        console.error(err);
    }
};
run();

// ====> needs to be bundled with Webpack <====
```

Logging middleware (build stage and finalizeRequest stage)

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/50976344/173452581-6157dfd5-5350-47a5-a1fd-33764e8212b3.png">



additionally, please take a look at this [issue](https://github.com/aws/aws-sdk-js-v3/issues/3610). Perhaps its related.

**Comment by trivikr** [comment]
I'm able to reproduce the issue using 😄 emoji only in AWS SDK for JavaScript (v3)

To reproduce create an example DynamoDB table, or using the code below:

<details>
<summary>createTable</summary>

```js
import { DynamoDB } from "@aws-sdk/client-dynamodb";

const TableName = "aws-sdk-js-v3-3559";

const params = {
  TableName,
  AttributeDefinitions: [{ AttributeName: "id", AttributeType: "S" }],
  KeySchema: [{ AttributeName: "id", KeyType: "HASH" }],
  BillingMode: "PAY_PER_REQUEST",
};

const region = "us-west-2";
const client = new DynamoDB({ region });
client.createTable(params);

```

</details>

Use the following patch file in https://github.com/aws-samples/aws-sdk-js-tests

[put-item-with-emoji.txt](https://github.com/aws/aws-sdk-js-v3/files/9378244/put-item-with-emoji.txt)

Code runs successfully in Node.js as follows:

```console
$ yarn start:node
...
Data returned by v2:
{
  "Item": {
    "id": {
      "S": "Emoji from v2 😄"
    }
  }
}

Data returned by v3:
{
  "$metadata": {
    "httpStatusCode": 200,
    "requestId": "B10RV813CNJFVV0EMVC8P987JRVV4KQNSO5AEMVJF66Q9ASUAAJG",
    "attempts": 1,
    "totalRetryDelay": 0
  },
  "Item": {
    "id": {
      "S": "Emoji from v3 😄"
    }
  }
}
```

Code fails in v3 on browser with the following error in browser console:

```console
The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.
```

<details>
<summary>Screenshot</summary>

![emoji-signature-error](https://user-images.githubusercontent.com/16024985/185513511-24323727-4c9a-4512-ab64-79d3ed141659.png)


</details>

**Comment by trivikr** [comment]
It appears to be an issue with Content-Length in the browser request.

Here is a patch file with a middleware which prints headers
[put-item-with-emoji-and-middleware.txt](https://github.com/aws/aws-sdk-js-v3/files/9378281/put-item-with-emoji-and-middleware.txt)

With the text `Hello World!`, here is the diff between Node.js and browser

<details>
<summary>string no emoji diff</summary>

![string-no-emoji-diff](https://user-images.githubusercontent.com/16024985/185515008-094b79be-ab1c-4a22-ab0c-e1d4c94d4370.png)

Rename `.txt` to `.html` to view after download: [string-no-emoji-diff.txt](https://github.com/aws/aws-sdk-js-v3/files/9378303/string-no-emoji-diff.txt)

</details>

With the text `Hello World!😄` the content-length in Node.js is 73, while that in browser is 75.

<details>
<summary>string with emoji diff</summary>

![string-emoji-diff](https://user-images.githubusercontent.com/16024985/185515362-7c6be27f-e67b-4c24-a521-0955352ee2a4.png)

Rename `.txt` to `.html` to view after download: [string-emoji-diff.txt](https://github.com/aws/aws-sdk-js-v3/files/9378310/string-emoji-diff.txt)

</details>

**Comment by trivikr** [comment]
Fix posted in https://github.com/aws/aws-sdk-js-v3/pull/3866
